### ISSUE TITLE
Ahead of Time Blade View Compilation

### ISSUE THREAD

**Issue by iamacarpet** [a1comms/GaeSupportLaravel#26]
Since App Engine is deploy once, run many times on many instances, there may be some performance benefit to ahead-of-time compilation of all of the blade views as deploy time.

This may not be standard in Laravel, as a deployment to a standard LAMP stack where the files reside on disk have the expectation that files can be changed and the framework will adapt dynamically.

We'll investigate if we can have an artisan command to do this as part of our Cloud Build push-to-deploy pipeline, creating the on-disk cache files that would then become read-only and should be valid for the lifetime of that version.

Once that's done, we should also look to see what the framework is doing to allow the files to be refreshed by invalidating the cache when the change on disk in the normal LAMP use case, as there may be some disk operations we can nullify here, hopefully reducing request latency where disk I/O latency is high on App Engine.

**Comment by iamacarpet** [comment]
Looks like this is possible.

To implement it, we probably want to copy `Illuminate\View\ViewServiceProvider` and provide an alternative version for use with pre-compiled views.

It should provide an alternate `CompilerEngine` that removes the following when running on GAE:
```php
        // If this given view has expired, which means it has simply been edited since
        // it was last compiled, we will re-compile the views so we can evaluate a
        // fresh copy of the view. We'll pass the compiler the path of the view.
        if ($this->compiler->isExpired($path)) {
            $this->compiler->compile($path);
        }
```
(don't forget to trace the new class and set an attribute precompiled=true).

It looks like there is also a `Illuminate\View\FileViewFinder` which scans the file-system for if views exist, so it may be worth generating a manifest (as a PHP file for opcache) in the compiled view directory, then lazy-loading in a custom `FileViewFinder`, so if it gets called it'll not have to do any file-system scans.

**Comment by iamacarpet** [comment]
This is now finished, works as expected.

It's integrated into the `gae:prepare` command, with automatic detection of if the pre-compiled view provider is loaded, to compile the views at deploy time.

After some basic testing, these are the latency results, taken using the trace functionality in this library, on the `laravel/view` span. I tested using a blank install of Laravel 5.5, just displaying the default `welcome.blade.php` view.

  | Standard | Pre-Compiled
-- | -- | --
1st Request | 5.661ms | 3.368ms
2nd Request | 1.153ms | 0.715ms
5th Request | 0.143ms | 0.054ms

On some of our more complex applications where there are hundreds of views with a lot of nested rendering, these small reductions might turn into quite a large latency saving.

For applications like the one we used for testing with just a basic one view per page, the saving may be negligible, but it will mean we are following the recommended workflow with Docker of doing all the generation at build time, then having everything run from the static build image (and not reading/writing/loading generated PHP files from /tmp).
